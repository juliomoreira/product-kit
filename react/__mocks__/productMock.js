export const productMock = itemNumber => {
  const itemMock = name => {
    return {
      benefitProduct: {
        linkText: 'linkText',
        productName: 'productName',
        brand: 'brand',
        items: [
          {
            productName: 'productName',
            itemId: 'itemId',
            name: name,
            image: [
              {
                imageId: 'imageId',
                imageUrl: '',
                imageText: 'imageText',
              },
            ],
            seller: [
              {
                sellerId: 'sellerId',
                commertialOffer: {
                  Installments: [
                    {
                      Value: 1,
                      InterestRate: 1,
                      NumberOfInstallments: 1,
                    },
                  ],
                  Price: 1,
                  ListPrice: 1,
                },
              },
            ],
          },
        ],
        skuId: 'skuId',
        slug: 'slug',
        productQuery: {
          product: {
            productId: 'productId',
            productName: 'productName',
            items: [
              {
                itemId: 'itemId',
                name: 'name',
                images: [
                  {
                    imageId: 'imageId',
                    imageUrl: '',
                    imageText: 'imageText',
                  },
                ],
                sellers: [
                  {
                    sellerId: 'sellerId',
                    commertialOffer: {
                      Installments: [
                        {
                          Value: 1,
                          InterestRate: 1,
                          NumberOfInstallments: 1,
                        },
                      ],
                      Price: 1,
                      ListPrice: 1,
                    },
                  },
                ],
              },
            ],
          },
          loading: false,
        },
        share: { social: {} },
      },
      benefitSKUIds: ['itemId'],
      discount: 1,
      minQuantity: 1,
    }
  }

  const mock = {
    productQuery: {
      loading: false,
      product: {
        benefits: [
          {
            items: [],
          },
        ],
        productId: '1',
      },
    },
  }

  let customNameCount = 0

  while (mock.productQuery.product.benefits[0].items.length < itemNumber) {
    customNameCount++
    mock.productQuery.product.benefits[0].items = mock.productQuery.product.benefits[0].items.concat(
      itemMock(`name-${customNameCount}`)
    )
  }

  return mock
}
