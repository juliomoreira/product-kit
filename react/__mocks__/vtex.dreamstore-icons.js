import React from 'react'

const iconMock = (orientation, size, className, name) => {
  return (
    <svg
      className={`${orientation}-oritentation-mock ${className} ${name}`}
      width={size}
      height={size}
    >
      <rect width={size} height={size} />
    </svg>
  )
}

export const IconEquals = ({ orientation, size, className }) =>
  iconMock(orientation, size, className, 'IconEquals')

export const IconPlus = ({ orientation, size, className }) =>
  iconMock(orientation, size, className, 'IconPlus')
