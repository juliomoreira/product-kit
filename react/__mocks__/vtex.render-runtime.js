import React from 'react'

export const ExtensionPoint = ({ id, product }) => {
  return <div className={`extension-point-${id}`}>{product.sku.name}</div>
}
