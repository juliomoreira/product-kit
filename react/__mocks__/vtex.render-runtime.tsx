import React from 'react'

export function ExtensionPoint() {
  return <div className="extension-point-mock" />
}
