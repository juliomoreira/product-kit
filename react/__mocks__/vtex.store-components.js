import React from 'react'

export const Slider = ({ children }) => {
  return <div className="slider-mock">{children}</div>
}

export const Container = ({ children }) => {
  return <section className="container-mock">{children}</section>
}

export const ProductPrice = ({ children }) => {
  return <div className="product-price-mock">{children}</div>
}

export const BuyButton = ({ children }) => {
  return <div className="buybutton-mock">{children}</div>
}
