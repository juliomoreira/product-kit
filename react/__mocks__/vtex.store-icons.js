import React from 'react'

const iconMock = (size, className, name) => {
  return (
    <svg className={`${className} ${name}`} width={size} height={size}>
      {name}
      <rect width={size} height={size} />
    </svg>
  )
}

export const IconEquals = ({ size, className }) =>
  iconMock(size, className, 'IconEquals')

export const IconPlus = ({ size, className }) =>
  iconMock(size, className, 'IconPlus')

export const IconRemove = ({ size, className }) =>
  iconMock(size, className, 'IconRemove')

export const IconSwap = ({ size, className }) =>
  iconMock(size, className, 'IconSwap')
