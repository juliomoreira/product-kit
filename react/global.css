.vtex-product-kit__details {
  max-width: 17.5em;
}

.vtex-product-kit__details .vtex-price-list__container--loader {
  height: 0;
}

.vtex-product-kit__details .vtex-price-installments--loader {
  height: 0;
}

.vtex-product-kit__details .vtex-price-selling--loader {
  height: 1.5em;
}

.vtex-product-kit__item-swap-button:hover div {
  display: inline;
}
