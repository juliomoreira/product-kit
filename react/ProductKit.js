import './global.css'

import React, { Component } from 'react'

import { path, equals } from 'ramda'
import { extractItemsKit } from './helpers'

import { getSchema } from './schema'
import { propTypes, defaultProps } from './props/productKitProps'

import ProductKitContent from './components/ProductKitContent'

const DEFAULT_MAX_VISIBLE_ITEMS = 3
const DEFAULT_VISIBLE_ITEMS = Array(DEFAULT_MAX_VISIBLE_ITEMS).fill(null)

/**
 * ProductKit component.
 * Wraps a ProductKitContent and manages de shown and hiden items kit.
 */
export default class ProductKit extends Component {
  static propTypes = propTypes

  static defaultProps = defaultProps

  static getSchema = getSchema

  state = {
    shownItems: DEFAULT_VISIBLE_ITEMS,
    hidenItems: [],
  }

  /**
   * Retrieve the items of a product.
   */
  getItems = product => {
    return path(['0', 'items'], path(['benefits'], product))
  }

  /**
   * When the component mounts it must update the shown and hiden items
   */
  componentDidMount() {
    this.updateComponentState(this.getItems(this.props.productQuery.product))
  }

  /**
   * Checks if the items data has changed since the last component
   * props update.
   */
  componentDidUpdate(prevProps) {
    const items = this.getItems(this.props.productQuery.product)
    const prevItems = this.getItems(prevProps.productQuery.product)

    if (
      equals(this.state.shownItems, DEFAULT_VISIBLE_ITEMS) ||
      !equals(items, prevItems)
    ) {
      this.updateComponentState(items)
    }
  }

  /**
   * Updates the shown and hiden items arrays with the content of the
   * items array passed as an argument. This function uses the helper
   * function to extract the items kit from the items data.
   */
  updateComponentState = items => {
    const itemsKit = extractItemsKit(items)

    if (itemsKit.length) {
      this.setState({
        shownItems: itemsKit.slice(0, DEFAULT_MAX_VISIBLE_ITEMS),
        hidenItems: itemsKit.slice(DEFAULT_MAX_VISIBLE_ITEMS),
      })
    }
  }

  /**
   * Receives the index of the item of the shownItems array and swap it by the
   * first item of the hidenItems array, the swapped item will be pushed at the end
   * of the hidenItems.
   */
  handleItemSwap = index => {
    const { shownItems, hidenItems } = this.state
    const item = shownItems[index]

    shownItems[index] = hidenItems.shift()
    hidenItems.push(item)

    this.setState({
      shownItems,
      hidenItems,
    })
  }

  render() {
    const {
      showBadge,
      badgeText,
      showLabels,
      showListPrice,
      showInstallments,
      productQuery: { product },
    } = this.props

    const benefits = path(['benefits'], product)

    /** The product does not have any Kit associated with it, in this case
     *  the ProductKitContent should not be rendered */
    if (benefits && !benefits.length) return null

    /** Shown and Hiden items */
    const { shownItems, hidenItems } = this.state

    /** The component is loading if the shown items has the default content */
    const loading = equals(shownItems, DEFAULT_VISIBLE_ITEMS)

    /** Allow item swap only if there's hiden items */
    const allowSwap = hidenItems.length > 0

    return (
      <div className="vtex-page-padding">
        <ProductKitContent
          loading={loading}
          allowSwap={allowSwap}
          itemsKit={shownItems}
          viewOptions={{
            showBadge,
            badgeText,
            showLabels,
            showListPrice,
            showInstallments,
          }}
          onItemSwap={this.handleItemSwap}
        />
      </div>
    )
  }
}
