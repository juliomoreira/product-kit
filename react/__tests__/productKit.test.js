import React from 'react'
import { render, fireEvent, act } from '@vtex/test-tools/react'
import { productMock } from '../__mocks__/productMock'
import ProductKit from '../index'

describe('<ProductSummary /> component', () => {
  function renderComponent(customProps, itemNumber = 1) {
    return render(<ProductKit {...productMock(itemNumber)} {...customProps} />)
  }

  it('should be rendered', () => {
    expect(renderComponent()).toBeDefined()
  })

  it('should match the snapshot with no removal and swap', () => {
    const { asFragment } = renderComponent({
      allowSwap: false,
      allowRemoval: false,
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('should match the snapshot with removal and swap', () => {
    const { asFragment } = renderComponent(
      {
        allowSwap: true,
        allowRemoval: true,
      },
      5
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render 3 products', () => {
    const { getByText } = renderComponent({}, 4)
    expect(getByText(/Take all 3 products/)).toBeTruthy()
  })

  it('should remove an item from the kit', () => {
    const { getByText, container } = renderComponent({ allowRemoval: true }, 2)
    expect(getByText(/Take all 2 products/)).toBeTruthy()

    act(() => {
      fireEvent.click(container.querySelector('.IconRemove'))
    })

    expect(getByText(/Take all 1 products/)).toBeTruthy()
  })

  /** It will swap the first swapable (name-2) for the one left over (name-4).
   * If swapped again, it should retrieve the previous one. */
  it('should swap an item from the kit', () => {
    const { getByText, container } = renderComponent({ allowSwap: true }, 4)
    expect(getByText('name-2')).toBeTruthy()

    act(() => {
      fireEvent.click(container.querySelector('.IconSwap'))
    })

    expect(getByText('name-4')).toBeTruthy()
    expect(getByText('name-2')).toBeFalsy()

    act(() => {
      fireEvent.click(container.querySelector('.IconSwap'))
    })

    expect(getByText('name-2')).toBeTruthy()
  })
})
