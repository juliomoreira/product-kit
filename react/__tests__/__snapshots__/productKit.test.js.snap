// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSummary /> component should match the snapshot 1`] = `
<DocumentFragment>
  <section
    class="container-mock"
  >
    <div
      class="t-heading-3 c-muted-1 fw3 mv4 flex items-center justify-center"
    >
      <span>
        Mount your Kit
      </span>
    </div>
    <div
      class="slider-mock"
    >
      <div
        class="container flex flex-row items-center justify-center b--muted-5 pv3 w-100"
      >
        <div
          class="flex flex-row items-center justify-center c-muted-1"
        >
          <div
            class="item"
          >
            <div
              class="flex flex-row"
            >
              <div
                class="pl4"
              >
                <div
                  class="extension-point-mock"
                />
              </div>
              <div
                class="flex flex-column items-center w2"
              />
            </div>
          </div>
        </div>
        <span
          class="c-muted-1"
        >
          <svg
            class="w2 IconEquals"
            height="44"
            width="44"
          >
            <rect
              height="44"
              width="44"
            />
          </svg>
        </span>
        <div
          class="listDetails flex flex-column items-center justify-center tc ph7"
        >
          <div
            class="t-body c-muted-1 mv3"
          >
            <span>
              Take all 1 products
            </span>
          </div>
          <div
            class="pv4"
          >
            <div
              class="product-price-mock"
            />
          </div>
          <div
            class="buybutton-mock"
          >
            <span>
              Buy Kit
            </span>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`<ProductSummary /> component should match the snapshot with removal and swap 1`] = `
<DocumentFragment>
  <section
    class="container-mock"
  >
    <div
      class="t-heading-3 c-muted-1 fw3 mv4 flex items-center justify-center"
    >
      <span>
        Mount your Kit
      </span>
    </div>
    <div
      class="slider-mock"
    >
      <div
        class="container flex flex-row items-center justify-center b--muted-5 pv3 w-100"
      >
        <div
          class="flex flex-row items-center justify-center c-muted-1"
        >
          <div
            class="item"
          >
            <div
              class="flex flex-row"
            >
              <div
                class="pl4"
              >
                <div
                  class="extension-point-mock"
                />
              </div>
              <div
                class="flex flex-column items-center w2"
              />
            </div>
          </div>
        </div>
        <div
          class="flex flex-row items-center justify-center c-muted-1"
        >
          <svg
            class="w2 IconPlus"
            height="45"
            width="45"
          >
            <rect
              height="45"
              width="45"
            />
          </svg>
          <div
            class="item"
          >
            <div
              class="flex flex-row"
            >
              <div
                class="pl4"
              >
                <div
                  class="extension-point-mock"
                />
              </div>
              <div
                class="flex flex-column items-center w2"
              >
                <div
                  class="pointer flex items-center mh3 mv3 c-on-base"
                >
                  <svg
                    class="w1 h1 IconRemove"
                    height="21"
                    width="21"
                  >
                    <rect
                      height="21"
                      width="21"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <span
          class="c-muted-1"
        >
          <svg
            class="w2 IconEquals"
            height="44"
            width="44"
          >
            <rect
              height="44"
              width="44"
            />
          </svg>
        </span>
        <div
          class="listDetails flex flex-column items-center justify-center tc ph7"
        >
          <div
            class="t-body c-muted-1 mv3"
          >
            <span>
              Take all 2 products
            </span>
          </div>
          <div
            class="pv4"
          >
            <div
              class="product-price-mock"
            />
          </div>
          <div
            class="buybutton-mock"
          >
            <span>
              Buy Kit
            </span>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
